### PR TITLE
API, Core: Support explicit null default values

### LINKED ISSUE #15932: `write-default` null representation issue
### Apache Iceberg version

None

### Query engine

None

### Please describe the bug 🐞

TL;DR:
It's not always possible to set the DEFAULT value for newly written rows to NULL, without also altering what the DEFAULT value for existing rows should be.

Because of the backwards-compatibility of `write-default`, it will default to `initial-default` if it's not present.
To represent `NULL`, the default has to be omitted.

`null` seems to not be an accepted state for either `initial-default` or `write-default`.
The spec doesn't say they're nullable, and the iceberg-rest-fixture Catalog will return 500 if it reads a `initial-default` or `write-default` that is `null`

That means it's impossible to set the `write-default` to NULL, unless `initial-default` is also NULL (omitted)

Table to illustrate the problem:
```
┌──────────────────┬────────────────┬────────────────────┐
│ initial-default  │ write-default  │ result (new rows)  │
├──────────────────┼────────────────┼────────────────────┤
│ 5                │ -              │ 5                  │
│ -                │ 21             │ 21                 │
│ 5                │ 21             │ 21                 │
│ -                │ -              │ NULL               │
└──────────────────┴────────────────┴────────────────────┘
```

`write-default` is omitted (`-`), but because `initial-default` is non-null, the result won't be NULL

### Willingness to contribute

- [ ] I can contribute a fix for this bug independently
- [ ] I would be willing to contribute a fix for this bug with guidance from the Iceberg community
- [x] I cannot contribute a fix for this bug at this time

### PR DESCRIPTION
**Pull request by huaxingao**
Fixes #15932.

Parsing "write-default": null in JSON crashes because Expressions.lit(null) rejects null. This makes it impossible to set write-default to null when initial-default is non-null — the fallback always wins.
This PR adds a NullLiteral sentinel (Literal.ofNull()) to distinguish "not set" from "explicitly null," and fixes the parser, serializer, and SchemaUpdate to handle it correctly.

Used Cursor - Claude-4.6-opus-high

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 1]
```diff
@@ -747,4 +752,50 @@ public String toString() {
       return "X'" + BASE16_ENCODING.encode(bytes) + "'";
     }
   }
+
+  /**
+   * A sentinel literal representing an explicit null default value. This is distinct from a Java
+   * {@code null} literal reference, which means "not set." Engines use this distinction to decide
+   * whether to fall back to initial-default when write-default is not set.
+   */
+  static class NullLiteral<T> implements Literal<T> {
```
Comment (anoopj) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:761]: `final`?

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 2]
```diff
@@ -880,12 +880,22 @@ private NestedField(
...
+          isOptional || this.writeDefault != Literal.ofNull(),
+          "Cannot use null write default for required field: %s",
+          name);
     }
 
     private static Literal<?> castDefault(Literal<?> defaultValue, Type type) {
       if (type.isNestedType() && defaultValue != null) {
         throw new IllegalArgumentException(
             String.format("Invalid default value for %s: %s (must be null)", type, defaultValue));
+      } else if (defaultValue == Literal.ofNull()) {
```
Comment (anoopj) [inline api/src/main/java/org/apache/iceberg/types/Types.java:897]: Would it make sense to do equals() instead of object identity?

**Review by anoopj — APPROVED** [review]
Change looks reasonable to me. Thank you! A couple of minor comments.

File: api/src/main/java/org/apache/iceberg/expressions/Literal.java [hunk 3]
```diff
@@ -71,6 +71,18 @@ static Literal<BigDecimal> of(BigDecimal value) {
     return new Literals.DecimalLiteral(value);
   }
 
+  /**
+   * Returns a sentinel literal representing an explicit null default value.
+   *
+   * <p>This is distinct from a Java {@code null} literal reference, which means "not set." When
```
Comment (ebyhr) [inline api/src/main/java/org/apache/iceberg/expressions/Literal.java:77]:
```suggestion
   * <p>This is distinct from a Java {@code null} literal reference, which means "not set". When
```

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 4]
```diff
@@ -747,4 +752,50 @@ public String toString() {
       return "X'" + BASE16_ENCODING.encode(bytes) + "'";
     }
   }
+
+  /**
+   * A sentinel literal representing an explicit null default value. This is distinct from a Java
+   * {@code null} literal reference, which means "not set." Engines use this distinction to decide
```
Comment (ebyhr) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:758]:
```suggestion
   * {@code null} literal reference, which means "not set". Engines use this distinction to decide
```

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 5]
```diff
@@ -880,12 +880,22 @@ private NestedField(
       this.doc = doc;
       this.initialDefault = castDefault(initialDefault, type);
       this.writeDefault = castDefault(writeDefault, type);
+      Preconditions.checkArgument(
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/types/Types.java]: think we should probably move this checkArg into "castDefault". That way we can tie the invariant to setting the default instead of doing it after the defaults are set.

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 6]
```diff
@@ -880,12 +880,22 @@ private NestedField(
...
+          "Cannot use null initial default for required field: %s",
+          name);
+      Preconditions.checkArgument(
+          isOptional || this.writeDefault != Literal.ofNull(),
+          "Cannot use null write default for required field: %s",
+          name);
     }
 
     private static Literal<?> castDefault(Literal<?> defaultValue, Type type) {
       if (type.isNestedType() && defaultValue != null) {
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/types/Types.java]: There is probably a bug here because we should be allowing a NestedType with defaultValue = Literal.ofNull.

I think it probably makes sense to rename this function and change it's beahvior a little

Maybe do something like

initDefault(isOptional, defaultValue, type) {
    // Handle Unset (return null)
    // Reject Literal Null and Required
    // Handle Literal null (return NullLiteral)
    // Attempt Cast
}

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 1 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java]: Added. Thanks

File: api/src/main/java/org/apache/iceberg/types/Types.java (hunk 2 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/types/Types.java]: Right. Fixed. Thanks!

File: api/src/main/java/org/apache/iceberg/expressions/Literal.java (hunk 3 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/expressions/Literal.java]: Done

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 4 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java]: Done

File: api/src/main/java/org/apache/iceberg/types/Types.java (hunk 5 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/types/Types.java]: Sounds good. Moved.

File: api/src/main/java/org/apache/iceberg/types/Types.java (hunk 6 above)
Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/types/Types.java]: Fixed. Thanks!

**Comment by huaxingao** [comment]
@RussellSpitzer Could you please take one more look when you have a moment? Thanks!

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 7]
```diff
@@ -878,22 +878,25 @@ private NestedField(
...
+      this.writeDefault = initDefault(isOptional, writeDefault, type);
     }
 
-    private static Literal<?> castDefault(Literal<?> defaultValue, Type type) {
-      if (type.isNestedType() && defaultValue != null) {
+    private static Literal<?> initDefault(boolean isOptional, Literal<?> defaultValue, Type type) {
+      if (defaultValue == null) {
+        return null;
+      } else if (defaultValue.equals(Literal.ofNull())) {
+        Preconditions.checkArgument(isOptional, "Cannot use null default for required field");
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/types/Types.java]: Let's see if we can give a bit more information here to the user. 
"Cannot use null default for required field %s"
```suggestion
            Preconditions.checkArgument(isOptional, "Cannot use null default for required field: %s (%s)", name, type);
```

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/types/Types.java]: Note the other error messages in this block are also not great, we should fix those ina followup

File: core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java [hunk 8]
```diff
@@ -284,6 +284,34 @@ public void testUpdateDefaultPreservesOtherMetadata() {
     assertThat(updated.asStruct()).isEqualTo(expected.asStruct());
   }
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java:286]: Needs a test that updating a schema with an explict literal null fails for required columns

File: core/src/test/java/org/apache/iceberg/TestSchemaParser.java [hunk 9]
```diff
@@ -155,4 +156,118 @@ public void testPrimitiveTypeDefaultValues(Type.PrimitiveType type, Literal<?> d
...
+            .withId(1)
+            .ofType(Types.IntegerType.get())
+            .withWriteDefault(Literal.ofNull())
+            .build();
+
+    assertThat(absent.writeDefaultLiteral()).isNull();
+    assertThat(explicitNull.writeDefaultLiteral()).isEqualTo(Literal.ofNull());
+    assertThat(absent).isNotEqualTo(explicitNull);
+  }
 }
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestSchemaParser.java:311]: Since we are allowing literal null for nested types we should have some tests for them as well here

File: api/src/main/java/org/apache/iceberg/types/Types.java (hunk 7 above)
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/types/Types.java]: Oh we don't have the name, I guess just keep the type for now and let's followup and improve this to also use the field name.

**Review by RussellSpitzer — APPROVED** [review]
I just have a few nits, but looks pretty good to me

File: api/src/main/java/org/apache/iceberg/expressions/Literal.java [hunk 10]
```diff
@@ -71,6 +71,18 @@ static Literal<BigDecimal> of(BigDecimal value) {
     return new Literals.DecimalLiteral(value);
   }
 
+  /**
+   * Returns a sentinel literal representing an explicit null default value.
```
Comment (singhpk234) [inline api/src/main/java/org/apache/iceberg/expressions/Literal.java:75]: What stops this from being used in Predicates, which are intentionally 2 valued, introducing this this way will make it prone to 3 valued boolean logic ? which i believe it will fail in the comparator later ?

Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/expressions/Literal.java:75]: When users build predicates, they normally do something like:
```
Expressions.equal("col", null)
```
or 
```
 Expressions.lit(null)
```
These will fail immediately with NullPointerException("Cannot create expression literal from null") because Literals.from(value) rejects null up front.
The only way a user might misuse this is by explicitly calling Literal.ofNull() and passing it into the low-level 
```
Expressions.predicate(EQ, "col", Literal.ofNull())
``` 
That's a rare path. If someone does take this path, it would compile but fail at runtime via NullLiteral.comparator() throwing UnsupportedOperationException with `NullLiteral has no comparator`.

Comment (huaxingao) [reply api/src/main/java/org/apache/iceberg/expressions/Literal.java:75]: added an explicit guard in `Expressions.predicate(Operation, String, Literal<T>)` that rejects NullLiteral at predicate construction `("Invalid expression literal: null, use isNull or notNull instead")`, plus a test.

**Comment by Tishj** [comment]
Please correct me if I'm wrong, but won't this also require changing the `rest-catalog-open-api.yaml` to make `null` an accepted value? To explicitly differentiate it from the omitted state?

**Comment by Tishj** [comment]
Bump
